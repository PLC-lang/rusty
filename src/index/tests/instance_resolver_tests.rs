use crate::{index::instance_iterator::Instance, test_utils::tests::index};

#[test]
fn global_vars_are_retrieved() {
    let (_, index) = index(
        "
    VAR_GLOBAL
    a,b : INT;
    ",
    );

    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn programs_are_retrieved() {
    let (_, index) = index(
        "
        PROGRAM MainProg
        END_PROGRAM
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn functions_are_not_retrieved() {
    let (_, index) = index(
        "
    FUNCTION main : DINT
    END_FUNCTION
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn functions_blocks_are_not_retrieved() {
    let (_, index) = index(
        "
    FUNCTION_BLOCK fb
    END_FUNCTION_BLOCK
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn program_variables_are_retrieved() {
    let (_, index) = index(
        "
        PROGRAM MainProg
        VAR
            a,b : DINT;
        END_VAR
        END_PROGRAM
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn global_struct_variables_are_retrieved() {
    let (_, index) = index(
        "
    TYPE str : STRUCT
        a,b : DINT;
    END_STRUCT
    END_TYPE
    VAR_GLOBAL
        gStr : str;
    END_VAR
    PROGRAM MainProg
    VAR
        pStr : str;
    END_VAR
    END_PROGRAM
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn nested_global_struct_variables_are_retrieved() {
    let (_, index) = index(
        "
    TYPE str : STRUCT
        a,b : str2;
    END_STRUCT
    END_TYPE
    TYPE str2 : STRUCT
        c,d : DINT;
    END_STRUCT
    END_TYPE
    VAR_GLOBAL
        gStr : str;
    END_VAR
    PROGRAM MainProg
    VAR
        pStr : str;
    END_VAR
    END_PROGRAM
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn global_fb_variables_are_retrieved() {
    let (_, index) = index(
        "
    FUNCTION_BLOCK fb
    VAR
        a,b : DINT;
    END_VAR
    END_FUNCTION_BLOCK
    VAR_GLOBAL
        gFb : fb;
    END_VAR
    PROGRAM MainProg
    VAR
        pFb : fb;
    END_VAR
    END_PROGRAM
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn array_instances_are_repeated() {
    let (_, index) = index(
        "
    FUNCTION_BLOCK fb
    VAR
        c,b : DINT;
    END_VAR
    END_FUNCTION_BLOCK
    PROGRAM MainProg
    VAR
        aFb : ARRAY[0..2] OF fb;
        aFb1 : ARRAY[0..2,0..1] OF fb;
        aFb3 : ARRAY[0..2] OF ARRAY[0..1] of fb;
    END_VAR
    END_PROGRAM
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn array_with_const_instances_are_repeated() {
    let (_, index) = index(
        "
    FUNCTION_BLOCK fb
    VAR
        c,b : DINT;
    END_VAR
    END_FUNCTION_BLOCK
    PROGRAM MainProg
    VAR CONSTANT
        size : DINT := 2;
    END_VAR
    VAR
        aFb : ARRAY[0..SIZE] OF fb;
    END_VAR
    END_PROGRAM
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn pointer_variables_are_not_retrieved() {
    let (_, index) = index(
        "
    FUNCTION_BLOCK fb
    VAR
        a,b : DINT;
    END_VAR
    END_FUNCTION_BLOCK
    PROGRAM MainProg
    VAR
        rFb : REF_TO fb;
    END_VAR
    END_PROGRAM
    ",
    );
    insta::assert_debug_snapshot!(index.find_instances().collect::<Vec<Instance<'_>>>());
}

#[test]
fn filter_on_variables_are_applied() {
    let (_, index) = index(
        "
    FUNCTION_BLOCK fb
    VAR
        a,b : DINT;
    END_VAR
    END_FUNCTION_BLOCK
    PROGRAM MainProg
    VAR
        rFb : fb;
    END_VAR
    END_PROGRAM
    VAR_GLOBAL CONSTANT
        gFb : fb;
    END_VAR
    ",
    );
    insta::assert_debug_snapshot!(index
        .filter_instances(|it, _| !it.is_constant())
        .collect::<Vec<Instance<'_>>>());
}
