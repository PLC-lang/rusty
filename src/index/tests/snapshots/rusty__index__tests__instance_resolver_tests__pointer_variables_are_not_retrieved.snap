---
source: src/index/tests/instance_resolver_tests.rs
expression: "index.find_instances().collect::<Vec<Instance<'_>>>()"
---
[
    (
        ExpressionPath {
            names: [
                Name(
                    "MainProg",
                ),
            ],
        },
        VariableIndexEntry {
            name: "MainProg_instance",
            qualified_name: "MainProg",
            initial_value: None,
            argument_type: ByVal(
                Global,
            ),
            is_constant: false,
            data_type_name: "MainProg",
            location_in_parent: 0,
            linkage: Internal,
            binding: None,
            source_location: SourceLocation {
                span: Range(
                    TextLocation {
                        line: 6,
                        column: 12,
                        offset: 98,
                    }..TextLocation {
                        line: 6,
                        column: 20,
                        offset: 106,
                    },
                ),
            },
            varargs: None,
        },
    ),
    (
        ExpressionPath {
            names: [
                Name(
                    "MainProg",
                ),
                Name(
                    "rFb",
                ),
            ],
        },
        VariableIndexEntry {
            name: "rFb",
            qualified_name: "MainProg.rFb",
            initial_value: None,
            argument_type: ByVal(
                Local,
            ),
            is_constant: false,
            data_type_name: "__MainProg_rFb",
            location_in_parent: 0,
            linkage: Internal,
            binding: None,
            source_location: SourceLocation {
                span: Range(
                    TextLocation {
                        line: 8,
                        column: 8,
                        offset: 123,
                    }..TextLocation {
                        line: 8,
                        column: 11,
                        offset: 126,
                    },
                ),
            },
            varargs: None,
        },
    ),
]
